Tests/Add-ProGetAsset.Tests.ps1

```diff
@@ -0,0 +1,188 @@
+& (Join-Path -Path $PSScriptRoot -ChildPath 'Initialize-Tests.ps1' -Resolve)
+
+function GivenSession 
+{
+    $script:session = New-ProGetTestSession
+    $feed = Test-ProGetFeed -Session $session -FeedName 'versions' -FeedType 'Asset'
+    if( !$feed )
+    {
+        New-ProGetFeed -Session $session -FeedName 'versions' -FeedType 'Asset'
+    }
+}
+
+function GivenAsset
+{
+    param(
+        [string]
+        $Name,
+        [string]
+        $directory,
+        [string]
+        $FilePath
+    )
+    $script:progetAssetName = $Name
+    new-item $FilePath -type 'file' -value 'test' -Force
+    $script:progetAssetDirectory = $directory
+    $script:filePath = $FilePath
+}
+
+function GivenAssetWithoutFile
+{
+    param(
+        [string]
+        $Name,
+        [string]
+        $directory,
+        [string]
+        $FilePath
+    )
+    $script:progetAssetName = $Name
+    $script:progetAssetDirectory = $directory
+    $script:filePath = $FilePath
+
+}
+function GivenSubDirectory
+{
+    param(
+        [string]
+        $Name
+    )
+    $script:directory = $Name
+    New-item -Path $name -type 'directory' -Force
+}
+function GivenAssetThatDoesntExist
+{
+    param(
+        [string]
+        $Name
+    )
+    $script:progetAssetName = $Name
+    $script:progetAssetDirectory = $null
+}
+
+function WhenAssetIsUploaded
+{
+    $Global:Error.Clear()
+    Add-ProGetAsset -Session $session -AssetName $progetAssetName -AssetDirectory $progetAssetDirectory -fileName $filePath -ErrorAction SilentlyContinue
+}
+
+function ThenDirectoryShouldBeCreated
+{
+    param(
+        [string]
+        $name
+    )
+    it ('should have created a asset directory named ''{0}''' -f $name) {
+        Test-ProGetFeed -Session $session -FeedName $name -FeedType 'Asset' | should -be $true
+    }
+}
+
+function ThenAssetShouldExist
+{
+    param(
+        [string]
+        $Name
+    )
+    it ('should contain the file {0}' -f $Name) {
+        Get-ProGetAsset -session $session -AssetDirectory $progetAssetDirectory | Where-Object { $_.name -match $name } | should -not -BeNullOrEmpty
+    }
+}
+
+function ThenAssetShouldNotExist
+{
+    param(
+        [string]
+        $Name
+    )
+    it ('should not contain the file {0}' -f $Name) {
+        Get-ProGetAsset -session $session -AssetDirectory $progetAssetDirectory | Where-Object { $_.name -match $name } | should -BeNullOrEmpty
+    }
+}
+
+function ThenErrorShouldBeThrown
+{
+    param(
+        [string]
+        $ExpectedError
+    )
+    It ('should write an error that matches ''{0}''' -f $ExpectedError) {
+        $Global:Error | Where-Object { $_ -match $ExpectedError } | Should -not -BeNullOrEmpty
+    }
+}
+
+function ThenNoErrorShouldBeThrown
+{
+    It 'should not write any errors' {
+        $Global:Error | Should -BeNullOrEmpty
+    }
+}
+
+function cleanup

```

Comment:
If a test fails, the `cleanup` function won't get called because the test immediately stops. Instead, change this to an `Init` function that runs before each test. Change all your `cleanup` methods to `Init` methods.

Reply:
I can do this for resetting the variables, but I cant really do this for removing the files and the assets until after the test. maybe add a try catch block to the `When` function that will cleanup if an error occurs?

Reply:
Done, used TestDrive for local files, and deleted Proget files at end of all tests in file